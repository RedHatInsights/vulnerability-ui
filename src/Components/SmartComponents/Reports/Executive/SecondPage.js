import React, { Fragment } from 'react';
import PropTypes from 'prop-types';

import { Section, Column, Chart, Table, Panel, PanelItem } from '@redhat-cloud-services/frontend-components-pdf-generator';
import { Text } from '@react-pdf/renderer';
import { truncate } from 'lodash';

import styles from '../Common/styles';
import messages from '../../../../Messages';

const SecondPage = ({ data, intl }) => {
    const cveDesc = (rule) => <Text style={styles.mb8}> {truncate(rule, { length: 300 })}</Text>;

    const {
        recentCvesData,
        CVSSChart,
        CVSSData,
        CVSSHeader,
        topCves
    } = data;

    return (
        <Fragment key="first-section">
            <Text style={styles.header}>{intl.formatMessage(messages.cvesHeader)}</Text>
            <Section title={intl.formatMessage(messages.executiveReportCvesByCVSS)}>
                <Column>
                    <Chart
                        chartType="pie"
                        subTitle="Cves"
                        title="100"
                        data={CVSSChart}
                    />
                </Column>
                <Column>
                    <Table
                        withHeader
                        rows={[CVSSHeader, ...CVSSData]}
                    />
                </Column>
            </Section>
            <Section title={intl.formatMessage(messages.executiveReportRecentlyPublishedCves)}>
                <Panel style={{ flex: 1 }}>
                    {
                        recentCvesData.map(([label, value])=> (
                            <Fragment key={label}>
                                <PanelItem style={{ flex: 1 }} title={label}>
                                    {value}
                                </PanelItem>
                            </Fragment>
                        ))
                    }
                </Panel>
            </Section>
            <Section title={intl.formatMessage(messages.executiveReportTop3)} withColumn={false}>
                {
                    topCves.map(cve => (
                        <Panel key={cve.synopsis} title={cve.synopsis} description={cveDesc(cve.description)}>
                            <PanelItem title={intl.formatMessage(messages.executiveReportPanelTitle)}>
                                {Number.parseFloat(cve.cvss3_score).toFixed(1)}
                            </PanelItem>
                            <PanelItem title={intl.formatMessage(messages.systemsExposed)}>
                                {cve.systems_affected.toString()}
                            </PanelItem>
                        </Panel>
                    ))

                }
            </Section>
        </Fragment>
    );
};

SecondPage.propTypes = {
    data: PropTypes.exact({
        recentCvesData: PropTypes.array,
        CVSSChart: PropTypes.array,
        CVSSData: PropTypes.array,
        CVSSHeader: PropTypes.array,
        topCves: PropTypes.array
    }),
    intl: PropTypes.any
};

export default SecondPage;
