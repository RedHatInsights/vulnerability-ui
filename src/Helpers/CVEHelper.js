import { Tooltip } from '@patternfly/react-core';
import { ExternalLinkAltIcon, ServerAltIcon } from '@patternfly/react-icons';
import { processDate } from '@redhat-cloud-services/frontend-components-utilities/files/cjs/helpers';
import React from 'react';
import { BUSINESS_RISK_OPTIONS, STATUS_OPTIONS } from './constants';
import { FormattedMessage } from 'react-intl';
import { InsightsSystemRule, InsightsNoSystemRule } from '../Components/SmartComponents/InsightsSystemRule/InsightsSystemRule';
import messages from '../Messages';

/**
 * Creates the tables rows for systems exposed table
 *
 * @param {Boolean} isLoading
 * @param {Array} payload fetched data
 * @param {Array} openedRows id of opened(expaned) row
 *
 */
export const createExposedSystemsTable = ({ isLoading, payload, openedRows }, cve) => {
    if (isLoading) {
        return { data: [], meta: {}, isLoading };
    }

    let { data, meta } = payload;

    const cvesCount = data && data.length;
    const tableRows = data.map(item => ({
        ...item.attributes,
        id: item.id,
        cve,
        patchAccess: meta.patch_access || false,
        isOpen: openedRows.includes(item.id),
        status: item.attributes.status_name,
        children: item.attributes.rule
            ? <InsightsSystemRule cve={cve} rule={item.attributes.rule} />
            : <InsightsNoSystemRule cve={cve}/>
    }));

    return { data: tableRows, meta: { ...meta, cvesCount  },  isLoading };

};

export function createSystemsListTable(affectedSystems) {
    let isLoading = affectedSystems.isLoading;
    if (!isLoading) {
        let payload = affectedSystems.payload.asMutable();
        let { data, meta } = payload;
        const dataWithStatus = data.map(item => ({
            ...item,
            attributes: {
                ...item.attributes,
                status: (
                    <span>
                        {item.attributes.cve_status_id !== item.attributes.status_id ? (
                            <Tooltip content={<FormattedMessage {...messages.onlyThisSystemCvePair} />}>
                                <ServerAltIcon />
                            </Tooltip>
                        ) : (
                            ''
                        )}{' '}
                        {item.attributes.status_id !== undefined &&
                            STATUS_OPTIONS.find(option => parseInt(option.value) === item.attributes.status_id).label}
                    </span>
                )
            }
        }));
        return { data: dataWithStatus, meta, isLoading };
    }

    return { data: (affectedSystems.payload && affectedSystems.payload.data) || [], meta: {}, isLoading };
}

export function createCveDetailsPage(cves) {
    let isLoading = cves.isLoading;
    let error = cves.error;
    if (!isLoading && !error) {
        let payload = cves.payload.asMutable();
        let { data, meta } = payload;

        data = {
            impact: data.attributes.impact,
            public_date: processDate(data.attributes.public_date),
            description: data.attributes.description,
            synopsis: data.attributes.synopsis,
            rh_link: createRHDBLink(data.attributes.synopsis, messages.RHDBLink),
            mitre_link: createMitreLink(data.attributes.synopsis),
            cvss3_score: data.attributes.cvss3_score,
            cvss2_score: data.attributes.cvss2_score,
            cvss2_metrics: data.attributes.cvss2_metrics,
            cvss3_metrics: data.attributes.cvss3_metrics,
            business_risk: BUSINESS_RISK_OPTIONS.find(
                option => option.value === data.attributes.business_risk_id.toString()
            ).label,
            business_risk_justification: data.attributes.business_risk_text,
            business_risk_id: data.attributes.business_risk_id,
            status: STATUS_OPTIONS.find(option => option.value === data.attributes.status_id.toString()).label,
            status_id: data.attributes.status_id,
            status_justification: data.attributes.status_text,
            systems_status_detail: data.attributes.systems_status_detail,
            systems_status_divergent: data.attributes.systems_status_divergent,
            rules: data.attributes.rules
        };

        return { data, meta, isLoading };
    }

    return { data: [], meta: {}, isLoading, error };
}

export function createRHDBLink(item, link = messages.RHDBLink, props) {
    return (
        <a target="_blank" rel="noopener noreferrer"  href={'https://access.redhat.com/security/cve/' + item} {...props}>
            {<FormattedMessage {...link} />}{<ExternalLinkAltIcon className="pf-u-ml-sm"/>}
        </a>
    );
}

export function createMitreLink(item) {
    return (
        <a target="_blank" rel="noopener noreferrer" href={'https://cve.mitre.org/cgi-bin/cvename.cgi?name=' + item}>
            {<FormattedMessage {...messages.MITRELink} />}{<ExternalLinkAltIcon />}
        </a>
    );
}

export const cveTableRowActions = (methods, rowIndex) => {
    return (rowIndex % 2 === 0) ?
        [
            {
                title: <FormattedMessage {...messages.editBusinessRisk} />,
                onClick: (event, rowId, rowData) =>
                    methods.showBusinessRiskModal([
                        {
                            id: rowData.id,
                            business_risk_id: rowData.business_risk_id,
                            justification: rowData.business_risk_justification
                        }
                    ])
            },
            {
                title: <FormattedMessage {...messages.editStatus} />,
                onClick: (event, rowId, rowData) =>
                    methods.showStatusModal([
                        {
                            id: rowData.id,
                            status_id: rowData.status_id,
                            justification: rowData.status_justification
                        }
                    ])
            }
        ] : null;
};

export const systemCveTableRowActions = (methods, entity, rowIndex) => {
    return (rowIndex % 2 === 0) ?
        [{
            title: <FormattedMessage {...messages.editStatus} />,
            onClick: (event, rowId, rowData) =>
                methods.showStatusModal([
                    {
                        id: rowData.id,
                        display_name: entity.display_name,
                        status_id: rowData.status_id,
                        cve_status_id: rowData.cve_status_id,
                        justification: rowData.status_justification,
                        cve_justification: rowData.cve_status_justification
                    }
                ],
                entity
                )
        }] : null;
};

export const systemExposedTableRowActions = (showStatusModal, cve, rowIndex) => {
    return (rowIndex % 2 === 0) ?
        [{
            title: <FormattedMessage {...messages.editStatus} />,
            onClick: (event, rowId, rowData) => {
                showStatusModal(
                    [{
                        id: cve.id,
                        status_id: cve.status_id,
                        justification: cve.justification
                    }],
                    [{
                        id: rowData.id,
                        display_name: rowData.display_name,
                        status_id: rowData.status_id,
                        justification: rowData.status_text
                    }]
                );
            }
        }] : null;
};

export const systemTableRowActions = (optOut, isOptOut) => [
    {
        title: (
            isOptOut
                ? <FormattedMessage {...messages.systemKebabResumeAnalysis} />
                : <FormattedMessage {...messages.systemKebabExcludeAnalysis} />
        ),
        onClick: (event, rowId, rowData) => {
            optOut([rowData.id]);
        }
    }
];
