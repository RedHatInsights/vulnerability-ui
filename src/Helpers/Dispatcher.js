import { addNotification } from '@red-hat-insights/insights-frontend-components/components/Notifications';
import ReducerRegistry from '../Utilities/ReducerRegistry';
export function dispatchAction(actionCreator) {
    const store = ReducerRegistry.getStore();
    return store.dispatch(actionCreator).catch(res => {
        if ('json' in res) {
            let parsedMessage = res.json();
            parsedMessage.then(arr => {
                for (let err of arr.errors) {
                    store.dispatch(
                        addNotification({
                            variant: 'danger',
                            title: 'Error ' + err.status,
                            description: err.detail
                        })
                    );
                }
            });
        } else {
            store.dispatch(
                addNotification({
                    variant: 'danger',
                    title: 'Error ' + res.status,
                    description: res.statusText
                })
            );
        }
    });
}
