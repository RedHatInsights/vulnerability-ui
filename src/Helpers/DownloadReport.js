import { downloadFile } from '@redhat-cloud-services/frontend-components-utilities/files/helpers';

class DownloadReport  {
    constructor() {
        this.defaultParams = {
            page: 1,
            page_size: Number.MAX_SAFE_INTEGER,
            date: new Date().toISOString().replace(/[T:]/g, '-').split('.')[0] + '-utc'
        };

        this.params = {
            cve: ''
        };

        this.pages = ['systems-exposed', 'system-cves', 'cves', 'system-list'];
        this.formats = ['csv', 'json'];
    }

    filename = (page) => {
        switch (page) {
            case 'systems-exposed':
                return `vulnerability_systems-exposed-${this.params.cve}--${this.defaultParams.date}`;
            case 'system-list':
                return `vulnerability_systems--${this.defaultParams.date}`;
            default:
                return `vulnerability_cves--${this.defaultParams.date}`;

        }
    }

    checkTypes = (fetchData, page, format) => {
        if (typeof fetchData !== 'function') {
            throw `exec expects the 1st param to be funtion. Received ${typeof fetchData}`;
        }

        if (!this.pages.includes(page)) {
            throw `Page ${page} does not exist`;
        }

        if (!this.formats.includes(format)) {
            throw `Format ${format} is not supported`;
        }
    }

    /**
     * Execute and download the report
     *
     * @param {Function} fetchData - Function to be called in order to fetch data.
     * @param {Object || Array} params - Object to be sent as request params. If you need to pass extra argument, pass an array.
     * @param {String} format - Supported formats to download are CSV and JSON.
     * @param {String} page - Name of the page based on which the filename will be determined.
     * @returns {void}
     */
    exec = async (fetchData, params, format, page) => {
        this.checkTypes(fetchData, page, format);
        this.params.cve = Array.isArray(params) ? params[0] : '';

        try {
            let { payload } = Array.isArray(params)
                ? await fetchData(this.params.cve, {
                    ...params[1],
                    ...this.defaultParams,
                    data_format: format
                }) :
                await fetchData({
                    ...params,
                    ...this.defaultParams,
                    data_format: format
                });

            let { data: response } = await payload;
            let data = format === 'json' ? JSON.stringify(response) : response;
            // downaloads the files
            downloadFile(data, this.filename(page), format);
        } catch (error) {
            throw `${error}`;
        }

    }

}

export default new DownloadReport();
