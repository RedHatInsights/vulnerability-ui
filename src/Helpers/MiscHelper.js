import { SecurityIcon, UnknownIcon } from '@patternfly/react-icons';
import { SortByDirection } from '@patternfly/react-table';
import findIndex from 'lodash/findIndex';
import propTypes from 'prop-types';
import React from 'react';
import { impactColorList, impactList, PUBLIC_DATE_OPTIONS } from './constants';
import qs from 'query-string';
import { IntlProvider } from '@redhat-cloud-services/frontend-components-translations';

export const dataShape = propTypes.shape({
    data: propTypes.oneOfType([propTypes.object, propTypes.array]),
    meta: propTypes.object,
    isLoading: propTypes.bool
});

/**
 * Based on the  allowedParams it will construct the arguments for the API call
 * which is defined in the vulnerabilities-client. vulnerabilities-client is written is typescript
 * so we need to define and pass every single parameter. Order in allowedParams is important
 *
 * @param {Object} apiProps
 * @param {Array} allowedParams - order should align with the params in the vulnerabilities-client
 */
export function constructParameters(apiProps, allowedParams) {
    if (apiProps) {
        Object.keys(apiProps).forEach(
            key => (
                apiProps[key] === undefined
                || apiProps[key] === ''
            )
            && delete apiProps[key]
        );

        const params = allowedParams.map(item => apiProps[item]);

        return params;
    }

    return [];
}

// TODO DRY:similar to constructParameters
export function constructURLParameters(urlParams, allowedParams) {
    if (urlParams) {
        const params = { ...urlParams };
        Object.keys(urlParams).forEach(
            key => (
                params[key] === undefined
                || params[key] === ''
                || !allowedParams.includes(key)
                || params[key] === false
            )
            && delete params[key]

        );
        return params;
    }
}

export function formatDate(date = new Date(), includeTime = false) {
    const prepend = (number) => `${`${number}`.length === 1 ? '0' : ''}${number}`;
    const toFormat = new Date(date);
    if (toFormat instanceof Date && !isNaN(toFormat)) {
        const year = toFormat.getFullYear();
        const month = toFormat.getMonth() + 1; // month is zero indexed
        const day = toFormat.getDate();
        const hour = toFormat.getUTCHours();
        const minute = toFormat.getUTCMinutes();

        return includeTime
            ? `${year}-${prepend(month)}-${prepend(day)} ${prepend(hour)}:${prepend(minute)} UTC`
            : `${year}-${prepend(month)}-${prepend(day)}`;
    }

    return toFormat;
}

/**
 * Contructs the filters which have more than one parameter that needs to be passed
 * and resets all the parameteres when it's needed
 *
 * @param {Object} filterParams
 * @return {Object}
 */
export function constructFilterParameters(filterParams) {
    if (Object.prototype.hasOwnProperty.call(filterParams, 'publish_date')) {
        let publicEntry = PUBLIC_DATE_OPTIONS.find(item => item.value === filterParams.publish_date);

        if (typeof publicEntry !== 'undefined') {
            filterParams.public_from = publicEntry.from && formatDate(publicEntry.from);
            filterParams.public_to = publicEntry.to && formatDate(publicEntry.to);
        } else {
            filterParams.public_from = undefined;
            filterParams.public_to = undefined;
            filterParams.publish_date = undefined;
        }
    }

    return filterParams;
}

export function getImpactDetails(impact) {
    if (Object.keys(impactList).includes(impact)) {
        return {
            title: impactList[impact].title,
            color: impactList[impact].color,
            text: impactList[impact].text,
            titleContent: (
                <SecurityIcon size="lg" />
            )
        };
    } else {
        return {
            title: 'Unknown',
            color: impactColorList.unknown,
            text: '',
            titleContent: (
                <UnknownIcon size="lg" />
            )
        };
    }
}

export const createSortBy = (header, value) => {
    if (value) {
        let direction = value.indexOf('-') === 0 ? SortByDirection.desc : SortByDirection.asc;
        value = value.replace(/^(-|\+)/, '');
        const index = findIndex(header, item => item.key === value);
        let sort = {
            index,
            direction
        };
        return sort;
    }

    return {};
};

export const handleSortColumn = (key, direction, header, currentSort, apply) => {
    let columnName = header[key].key;
    const useDefault = currentSort && currentSort.substr(1) !== columnName;
    if (direction === SortByDirection.desc || useDefault) {
        columnName = '-' + columnName;
    }

    apply({ sort: columnName });
};

export const updateStateSet = (stateSet, names, value) => {
    [].concat(names).forEach(name => {
        value ? stateSet.add(name) : stateSet.delete(name);
    });
    return stateSet;
};

export const useUrlParams = (allowedParams) => {
    const url = new URL(window.location);
    const urlParams = qs.parse(url.search);

    const setUrlParams = (parameters) => {
        const searchParams = qs.stringify(constructURLParameters(parameters, allowedParams));
        window.history.replaceState(null, null, `${url.origin}${url.pathname}?${searchParams}`);
    };

    return [urlParams, setUrlParams];
};

export const updateRef = (meta, apply) => {
    const pages = parseInt(meta.pages);
    const page = parseInt(meta.page);
    const cvesCount = parseInt(meta.cvesCount);
    const goTo = (pages === page && page > 1 && cvesCount === 1) ? (pages - 1) : page;

    apply({ page: goTo });
};

export const mountWithIntl = (Component) => {
    const wrapper = mount(Component, {
        wrappingComponent: IntlProvider
    });
    const provider = wrapper.getWrappingComponent();
    provider.setProps({ locale: 'en' });

    return wrapper;
};

export const shallowWithIntl = (Component) => {
    const wrapper = shallow(Component, {
        wrappingComponent: IntlProvider
    });
    const provider = wrapper.getWrappingComponent();
    provider.setProps({ locale: 'en' });

    return wrapper;
};

export const insertIf = (condition, ...elements) => {
    return condition ? elements : [];
};
