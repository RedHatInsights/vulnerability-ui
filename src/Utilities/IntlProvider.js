import { createIntl, createIntlCache } from 'react-intl';

const cache = createIntlCache();
const locale = navigator.language.slice(0, 2);

export const intl = createIntl({
    locale
}, cache);

export const intlSettings = { locale };
