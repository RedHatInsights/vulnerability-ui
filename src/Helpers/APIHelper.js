let BASE_ROUTE = '/api/vulnerability/v1';
import { GitApi } from '@redhat-cloud-services/vulnerabilities-client';
import { SystemProfileApi } from '@redhat-cloud-services/host-inventory-client';
import instance from '../Utilities/interceptors';
import { constructParameters } from './MiscHelper';

let api = new GitApi(undefined, BASE_ROUTE, instance);

const INVENTORY_API_BASE = '/api/inventory/v1';
const systemProfile = new SystemProfileApi(undefined, INVENTORY_API_BASE, instance);

const defaultParams = [
    'filter',
    'limit',
    'offset',
    'page',
    'page_size',
    'sort'
];

export function getCveDetails(synopsis) {
    const result = api.getCveDetails(synopsis);
    return result;
}

export function getSystemDetails(system) {
    const result = api.getSystemDetails(system);
    return result;
}

export function setSystemCveStatus(params) {
    let result = api.setStatus(params);
    return result;
}

export function optOutSystems(systems, optOut) {
    const data = { opt_out: optOut, inventory_id: new Array(...systems) };
    let result = api.setSystemsOptOut(data);
    return result;
}

export function setBusinessRisk(params) {
    let result = api.setCveBusinessRisk(params);
    return result;
}

export function setCveStatus(params) {
    let result = api.setCveStatus(params);
    return result;
}

/**
 * /vulnerabilities/cves
 * /vulnerabilities/cves/ids
 */
let cveParams = [
    'cvss_from',
    'cvss_to',
    'public_from',
    'public_to',
    'impact',
    'data_format',
    'business_risk_id',
    'status_id',
    'rule_presence',
    'tags',
    'sap_sids',
    'sap_system',
    'known_exploit',
    'affecting',
    'rhel_version'
];

export function getCveListByAccount(apiProps) {
    let parameterNames = [
        ...defaultParams,
        ...cveParams,
        'report',
        'advanced_report',
        'ansible',
        'mssql'
    ];
    let parameterArray = constructParameters(apiProps, parameterNames);
    let result = api.getCveList(...parameterArray);
    return result;
}

export function getCveIdsList(apiProps) {
    let parameterNames = [
        ...defaultParams,
        ...cveParams,
        'ansible',
        'mssql'
    ];

    let parameterArray = constructParameters(apiProps, parameterNames);
    let result = api.getCveIdsList(...parameterArray);
    return result;
}

/**
 * /cves/ID/affected_systems
 */
let systemsByCVEparams = [
    'status_id',
    'data_format',
    'uuid',
    'rule_key',
    'rule_presence',
    'rule',
    'tags',
    'sap_sids',
    'sap_system',
    'show_advisories',
    'advisory',
    'rhel_version',
    'first_reported_from',
    'first_reported_to',
    'advisory_available',
    'remediation'
];

export function getAffectedSystemsByCVE({ id, ...apiProps }) {
    let parameterNames = [
        ...defaultParams,
        ...systemsByCVEparams,
        'report',
        'ansible',
        'mssql'
    ];

    let parameterArray = constructParameters(apiProps, parameterNames);
    let result = api.getAffectedSystemsByCve(id, ...parameterArray);
    return result;
}

export function getAffectedSystemsIdsByCve({ id, ...apiProps }) {
    let parameterNames = [
        ...defaultParams,
        ...systemsByCVEparams,
        'ansible',
        'mssql'
    ];

    let parameterArray = constructParameters(apiProps, parameterNames);
    let result = api.getAffectedSystemsIdsByCve(id, ...parameterArray);
    return result;
}

/**
 * /systems/ID/cves
 * /systems/ID/cves/ids
 */
let cveBySystemParams = [
    'cvss_from',
    'cvss_to',
    'public_from',
    'public_to',
    'impact',
    'status_id',
    'data_format',
    'business_risk_id',
    'rule_presence',
    'show_advisories',
    'advisory',
    'rule_key',
    'known_exploit',
    'first_reported_from',
    'first_reported_to',
    'advisory_available',
    'remediation'
];

export function getCveListBySystem(apiProps) {
    const { system } = apiProps;
    let parameterNames = [
        ...defaultParams,
        ...cveBySystemParams,
        'report'
    ];

    if (apiProps && system) {
        Object.keys(apiProps).forEach(key => (apiProps[key] === undefined || apiProps[key] === '') && delete apiProps[key]);
        const params = parameterNames.map(item => apiProps[item]);
        return api.getCveListBySystem(system, ...params).catch(err => {
            if (err && err.status === '404') {
                return { errors: err };
            }

            throw err;
        });
    }
}

export function getCveIdsBySystem(apiProps) {
    const { system } = apiProps;
    let parameterNames = [
        ...defaultParams,
        ...cveBySystemParams
    ];

    if (apiProps && system) {
        Object.keys(apiProps).forEach(key => (apiProps[key] === undefined || apiProps[key] === '') && delete apiProps[key]);
        const params = parameterNames.map(item => apiProps[item]);
        return api.getCveIdsBySystem(system, ...params).catch(err => {
            if (err && err.status === '404') {
                return { errors: err };
            }

            throw err;
        });
    }
}

/**
 * /systems/ids
 * /systems
 */
let systemsParams = [
    'data_format',
    'stale',
    'uuid',
    'tags',
    'sap_sids',
    'sap_system',
    'excluded',
    'rhel_version'
];

export function getSystems(apiProps) {
    let parameterNames = [
        ...defaultParams,
        ...systemsParams,
        'report',
        'ansible',
        'mssql'
    ];
    let parameterArray = constructParameters(apiProps, parameterNames);
    let result = api.getSystemsList(...parameterArray);
    return result;
}

export function getSystemsIds(apiProps) {
    let parameterNames = [
        ...defaultParams,
        ...systemsParams
    ];
    let parameterArray = constructParameters(apiProps, parameterNames);
    let result = api.getSystemsIds(...parameterArray);
    return result;
}

export function getAnnouncement() {
    let result = api.getAnnouncement();
    return result;
}

export function getDashbar(apiProps) {
    let parameterNames = [
        'tags',
        'sap_sids',
        'sap_system',
        'ansible',
        'mssql'
    ];
    let parameterArray = constructParameters(apiProps, parameterNames);
    let result = api.getDashbar(...parameterArray);
    return result;
}

export function getOperatingSystems() {
    return systemProfile.apiSystemProfileGetOperatingSystem();
}

export function fetchPdfReport(template, filters) {
    const CRC_PDF_GENERATE_API = '/api/crc-pdf-generator/v1/generate';
    const url = new URL(CRC_PDF_GENERATE_API,  window.location.origin);

    return fetch(url, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json'
        },

        body: JSON.stringify({
            service: 'vulnerability',
            template,
            params: {
                ...filters
            }
        })
    }, 50000)
        .then((response) => response.blob());
}
