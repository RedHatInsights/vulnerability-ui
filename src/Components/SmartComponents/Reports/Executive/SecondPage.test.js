import SecondPage from './SecondPage';
import toJson from 'enzyme-to-json';
import { mountWithIntl } from '../../../../Helpers/MiscHelper';
import { BrowserRouter as Router } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { intl } from '../../../../Utilities/IntlProvider';

const mockStore = configureStore([store => next => action => {}]);
let store = mockStore({ });

const data = {
    "recentCvesData": [
        [
            "Last 7 days",
            "0"
        ],
        [
            "Last 30 days",
            "2"
        ],
        [
            "Last 90 days",
            "22"
        ]
    ],
    "CVSSChart": [
        {
            "x": "8.0 - 10.0",
            "y": 18
        },
        {
            "x": "4.0 - 7.9",
            "y": 75
        },
        {
            "x": "0.0 - 3.9",
            "y": 7
        }
    ],
    "CVSSData": [
        [
            "44 (7% of total)"
        ],
        [
            "446 (75% of total)"
        ],
        [
            "105 (18% of total)"
        ]
    ],
    "CVSSHeader": [
        "CVSS score range",
        "Number of CVEs"
    ],
    "topCves": [
        {
            "cvss2_score": "None",
            "cvss3_score": "8.200",
            "description": "A flaw was found in grub2, prior to version 2.06. An attacker may use the GRUB 2 flaw to hijack and tamper the GRUB  verification process. This flaw also allows the bypass of Secure Boot protections. In order to load an untrusted or modified kernel, an attacker would first need to establish access to the system such as gaining physical access, obtain the ability to alter a pxe-boot network, or have remote access to a  networked system with root access. With this access, an attacker could then craft a string to cause a buffer overflow by injecting a malicious payload that leads to arbitrary code execution within GRUB. The highest threat from this vulnerability is to data confidentiality and integrity as well as system availability.",
            "known_exploit": false,
            "rule_presence": true,
            "synopsis": "CVE-2020-10713",
            "systems_affected": 36
        },
        {
            "cvss2_score": "None",
            "cvss3_score": "8.600",
            "description": "Heap buffer overflow in Freetype in Google Chrome prior to 86.0.4240.111 allowed a remote attacker to potentially exploit heap corruption via a crafted HTML page.",
            "known_exploit": false,
            "rule_presence": false,
            "synopsis": "CVE-2020-15999",
            "systems_affected": 35
        },
        {
            "cvss2_score": "None",
            "cvss3_score": "8.100",
            "description": "A flaw was found in cloud-init, where it uses the random.choice function when creating sensitive random strings used for generating a random password in new instances. Depending on the instance configuration, a remote or local attacker may abuse this vulnerability to guess the password of the victim user.",
            "known_exploit": false,
            "rule_presence": false,
            "synopsis": "CVE-2020-8631",
            "systems_affected": 34
        }
    ]
}

describe('SecondPage component', () => {
    it('Should match snapshot', () => {
        const wrapper = mountWithIntl(
            <Provider store={store}>
                <Router>
                    <SecondPage data={data} intl={intl} />
                </Router>
            </Provider>);

        expect(toJson(wrapper)).toMatchSnapshot();
    })
});
