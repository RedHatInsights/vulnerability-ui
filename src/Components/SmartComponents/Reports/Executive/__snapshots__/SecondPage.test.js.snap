// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SecondPage component Should match snapshot 1`] = `
<Provider
  store={
    {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      location={
        {
          "hash": "",
          "key": "default",
          "pathname": "/",
          "search": "",
          "state": null,
        }
      }
      navigationType="POP"
      navigator={
        {
          "action": "POP",
          "createHref": [Function],
          "createURL": [Function],
          "encodeLocation": [Function],
          "go": [Function],
          "listen": [Function],
          "location": {
            "hash": "",
            "key": "default",
            "pathname": "/",
            "search": "",
            "state": null,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <SecondPage
        data={
          {
            "CVSSChart": [
              {
                "x": "8.0 - 10.0",
                "y": 18,
              },
              {
                "x": "4.0 - 7.9",
                "y": 75,
              },
              {
                "x": "0.0 - 3.9",
                "y": 7,
              },
            ],
            "CVSSData": [
              [
                "44 (7% of total)",
              ],
              [
                "446 (75% of total)",
              ],
              [
                "105 (18% of total)",
              ],
            ],
            "CVSSHeader": [
              "CVSS score range",
              "Number of CVEs",
            ],
            "recentCvesData": [
              [
                "Last 7 days",
                "0",
              ],
              [
                "Last 30 days",
                "2",
              ],
              [
                "Last 90 days",
                "22",
              ],
            ],
            "topCves": [
              {
                "cvss2_score": "None",
                "cvss3_score": "8.200",
                "description": "A flaw was found in grub2, prior to version 2.06. An attacker may use the GRUB 2 flaw to hijack and tamper the GRUB  verification process. This flaw also allows the bypass of Secure Boot protections. In order to load an untrusted or modified kernel, an attacker would first need to establish access to the system such as gaining physical access, obtain the ability to alter a pxe-boot network, or have remote access to a  networked system with root access. With this access, an attacker could then craft a string to cause a buffer overflow by injecting a malicious payload that leads to arbitrary code execution within GRUB. The highest threat from this vulnerability is to data confidentiality and integrity as well as system availability.",
                "known_exploit": false,
                "rule_presence": true,
                "synopsis": "CVE-2020-10713",
                "systems_affected": 36,
              },
              {
                "cvss2_score": "None",
                "cvss3_score": "8.600",
                "description": "Heap buffer overflow in Freetype in Google Chrome prior to 86.0.4240.111 allowed a remote attacker to potentially exploit heap corruption via a crafted HTML page.",
                "known_exploit": false,
                "rule_presence": false,
                "synopsis": "CVE-2020-15999",
                "systems_affected": 35,
              },
              {
                "cvss2_score": "None",
                "cvss3_score": "8.100",
                "description": "A flaw was found in cloud-init, where it uses the random.choice function when creating sensitive random strings used for generating a random password in new instances. Depending on the instance configuration, a remote or local attacker may abuse this vulnerability to guess the password of the victim user.",
                "known_exploit": false,
                "rule_presence": false,
                "synopsis": "CVE-2020-8631",
                "systems_affected": 34,
              },
            ],
          }
        }
        intl={
          {
            "$t": [Function],
            "defaultFormats": {},
            "defaultLocale": "en",
            "defaultRichTextElements": undefined,
            "fallbackOnEmptyString": true,
            "formatDate": [Function],
            "formatDateTimeRange": [Function],
            "formatDateToParts": [Function],
            "formatDisplayName": [Function],
            "formatList": [Function],
            "formatListToParts": [Function],
            "formatMessage": [Function],
            "formatNumber": [Function],
            "formatNumberToParts": [Function],
            "formatPlural": [Function],
            "formatRelativeTime": [Function],
            "formatTime": [Function],
            "formatTimeToParts": [Function],
            "formats": {},
            "formatters": {
              "getDateTimeFormat": [Function],
              "getDisplayNames": [Function],
              "getListFormat": [Function],
              "getMessageFormat": [Function],
              "getNumberFormat": [Function],
              "getPluralRules": [Function],
              "getRelativeTimeFormat": [Function],
            },
            "locale": "en",
            "messages": {},
            "onError": [Function],
            "onWarn": [Function],
            "textComponent": Symbol(react.fragment),
            "timeZone": undefined,
          }
        }
      >
        <TEXT
          style={
            {
              "color": "#C9190B",
              "fontSize": 20,
            }
          }
        >
          CVEs
        </TEXT>
        <Section
          title="Identified  CVEs by CVSS score"
          titleProps={
            {
              "style": {
                "fontSize": 10,
              },
            }
          }
        >
          <div />
        </Section>
        <Section
          title="Recently published CVEs indentified on systems"
          titleProps={
            {
              "style": {
                "fontSize": 10,
              },
            }
          }
        >
          <div />
        </Section>
        <Section
          title="Top 3 vulnerabilities in your infrastructure"
          titleProps={
            {
              "style": {
                "fontSize": 10,
              },
            }
          }
          withColumn={false}
        >
          <div />
        </Section>
      </SecondPage>
    </Router>
  </BrowserRouter>
</Provider>
`;
